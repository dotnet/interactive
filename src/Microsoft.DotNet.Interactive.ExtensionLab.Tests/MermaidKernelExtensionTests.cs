// Copyright (c) .NET Foundation and contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;

using Assent;

using FluentAssertions;
using FluentAssertions.Execution;
using HtmlAgilityPack;
using Microsoft.DotNet.Interactive.Commands;
using Microsoft.DotNet.Interactive.CSharp;
using Microsoft.DotNet.Interactive.Events;
using Microsoft.DotNet.Interactive.Formatting;
using Microsoft.DotNet.Interactive.Tests.Utility;

using Xunit;

namespace Microsoft.DotNet.Interactive.ExtensionLab.Tests
{
    public class MermaidKernelExtensionTests : IDisposable
    {
        private readonly Configuration _configuration;

        public MermaidKernelExtensionTests()
        {
            _configuration = new Configuration()
                .UsingExtension("txt")
                .SetInteractive(Debugger.IsAttached);
        }

        [Fact]
        public async Task adds_mermaid_kernel()
        {
            using CompositeKernel kernel = new CompositeKernel
            {
                new CSharpKernel().UseNugetDirective(),
            };

            var extension = new MermaidKernelExtension();

            await extension.OnLoadAsync(kernel);

            kernel.ChildKernels
                .Should()
                .ContainSingle(k => k is MermaidKernel);
        }

        [Fact]
        public async Task registers_html_formatter_for_MermaidMarkdown()
        {
            using CompositeKernel kernel = new CompositeKernel
            {
                new CSharpKernel().UseNugetDirective(),
            };

            var extension = new MermaidKernelExtension();

            await extension.OnLoadAsync(kernel);

            var markdown = @"graph TD
    A[Client] --> B[Load Balancer]
    B --> C[Server1]
    B --> D[Server2]";

            var formatted = new MermaidMarkdown(markdown).ToDisplayString(HtmlFormatter.MimeType);
            var doc = new HtmlDocument();
            doc.LoadHtml(formatted.FixedGuid());
            var scriptNode = doc.DocumentNode.SelectSingleNode("//div/script");
            var renderTarget = doc.DocumentNode.SelectSingleNode("//div[@id='00000000000000000000000000000000']");
            using var _ = new AssertionScope();

            scriptNode.Should().NotBeNull();
            scriptNode.InnerText.Should()
                .Contain(markdown);
            scriptNode.InnerText.Should()
                .Contain("configureRequireFromExtension('Mermaid','1.0.0')(['Mermaid/mermaidapi'], (mermaid) => {");

            renderTarget.Should().NotBeNull();
        }

        [Fact]
        public async Task registers_html_formatter_for_explorer()
        {
            using CompositeKernel kernel = new CompositeKernel
            {
                new CSharpKernel().UseNugetDirective(),
            };

            var extension = new MermaidKernelExtension();

            await extension.OnLoadAsync(kernel);

            var explorer = typeof(List<string>).ExploreWithUmlClassDiagram();
            var formatted = explorer.ToDisplayString(HtmlFormatter.MimeType);
            var markdown = explorer.ToMarkdown().ToString();

            var doc = new HtmlDocument();
            doc.LoadHtml(formatted.FixedGuid());
            var scriptNode = doc.DocumentNode.SelectSingleNode("//div/script");
            var renderTarget = doc.DocumentNode.SelectSingleNode("//div[@id='00000000000000000000000000000000']");
            
            using var _ = new AssertionScope();

            scriptNode.Should().NotBeNull();
            scriptNode.InnerText.Should()
                .Contain(markdown);
            scriptNode.InnerText.Should()
                .Contain("configureRequireFromExtension('Mermaid','1.0.0')(['Mermaid/mermaidapi'], (mermaid) => {");

            renderTarget.Should().NotBeNull();
        }


        [Fact]
        public async Task mermaid_kernel_handles_SubmitCode()
        {
            using CompositeKernel kernel = new CompositeKernel
            {
                new CSharpKernel().UseNugetDirective(),
            };

            var extension = new MermaidKernelExtension();

            await extension.OnLoadAsync(kernel);

            KernelCommandResult result = await kernel.SubmitCodeAsync(@"
#!mermaid
    graph TD
    A[Client] --> B[Load Balancer]
    B --> C[Server1]
    B --> D[Server2]
");

            var events = result.KernelEvents.ToSubscribedList();

            var formattedData = events
                .OfType<DisplayedValueProduced>()
                .Single()
                .FormattedValues
                .Single(fm => fm.MimeType == HtmlFormatter.MimeType)
                .Value;

            this.Assent(formattedData.FixedGuid(), _configuration);
        }

        [Fact]
        public async Task can_use_extension_methods_from_the_kernel_extension()
        {
            using CompositeKernel kernel = new CompositeKernel
            {
                new CSharpKernel().UseNugetDirective(),
            };

            var extension = new MermaidKernelExtension();

            await extension.OnLoadAsync(kernel);

            await kernel.SendAsync(new SubmitCode($@"#r ""{typeof(MermaidKernelExtension).Assembly.Location}""", "csharp"));

            var result = await kernel.SendAsync( new SubmitCode(@"
using System;

typeof(List<string>).ExploreWithUmlClassDiagram().Display();
", "csharp"));

            var events = result.KernelEvents.ToSubscribedList();

            var formattedData = events
                .OfType<DisplayedValueProduced>()
                .Single()
                .FormattedValues
                .Single(fm => fm.MimeType == HtmlFormatter.MimeType)
                .Value;

            this.Assent(formattedData.FixedGuid(), _configuration);
        }

        [Fact]
        public void can_generate_class_diagram_from_type()
        {
            var diagram = typeof(JsonElement).ToClassDiagram();

            this.Assent(diagram.ToString());
        }

        [Fact]
        public void can_generate_class_diagram_from_generic_type()
        {
            var diagram = typeof(List<Dictionary<string,object>>).ToClassDiagram();

            this.Assent(diagram.ToString());
        }


        [Fact]
        public void can_generate_class_diagram_to_a_specified_depth()
        {
            var diagram = typeof(List<Dictionary<string, object>>).ToClassDiagram(new ClassDiagramConfiguration(1));

            diagram.ToString().Should()
                .NotContain(
                    "ICollection~Dictionary<String, Object>~ --|> IEnumerable~Dictionary<String, Object>~ : Inheritance");
        }

        [Fact]
        public void can_explore_a_type_with_UmlClassDiagram()
        {
            var diagram = typeof(List<Dictionary<string, object>>).ExploreWithUmlClassDiagram().ToMarkdown();

            this.Assent(diagram.ToString());
        }

        [Fact]
        public void can_configure_UmlClassDiagramExplorer()
        {
            var diagram = typeof(List<Dictionary<string, object>>).ExploreWithUmlClassDiagram()
                .WithGraphDepth(1).ToMarkdown();
            
            diagram.ToString().Should()
                .NotContain(
                    "ICollection~Dictionary<String, Object>~ --|> IEnumerable~Dictionary<String, Object>~ : Inheritance");
        }

        public void Dispose()
        {
            Formatter.ResetToDefault();
        }
    }
}