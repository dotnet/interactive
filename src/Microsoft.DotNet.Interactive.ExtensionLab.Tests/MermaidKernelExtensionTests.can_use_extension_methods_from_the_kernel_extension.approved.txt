<div style="background-color:white;">
<script type="text/javascript">

loadMermaid_00000000000000000000000000000000 = () => {
 
        (require.config({ 'paths': { 'context': '1.0.252001', 'mermaidUri' : 'https://colombod.github.io/dotnet-interactive-cdn/extensionlab/1.0.252001/mermaid/mermaidapi', 'urlArgs': 'cacheBuster=00000000000000000000000000000000' }}) || require)(['mermaidUri'], (mermaid) => {

            let renderTarget = document.getElementById('00000000000000000000000000000000');
            mermaid.render( 
                'mermaid_00000000000000000000000000000000', 
                `classDiagram
List~String~ --|> ICollection~String~ : Inheritance
List~String~ --|> IEnumerable~String~ : Inheritance
List~String~ --|> IList~String~ : Inheritance
List~String~ --|> IReadOnlyCollection~String~ : Inheritance
List~String~ --|> IReadOnlyList~String~ : Inheritance
List~String~ --|> ICollection : Inheritance
List~String~ --|> IEnumerable : Inheritance
List~String~ --|> IList : Inheritance
List~String~ --|> Object : Inheritance

class ICollection
<<interface>> ICollection
ICollection : +CopyTo(Array array, Int32 index) Void*
ICollection : +get_Count() Int32*
ICollection : +get_IsSynchronized() Boolean*
ICollection : +get_SyncRoot() Object*

class ICollection~String~
<<interface>> ICollection~String~
ICollection~String~ : +Add(String item) Void*
ICollection~String~ : +Clear() Void*
ICollection~String~ : +Contains(String item) Boolean*
ICollection~String~ : +CopyTo(String[] array, Int32 arrayIndex) Void*
ICollection~String~ : +get_Count() Int32*
ICollection~String~ : +get_IsReadOnly() Boolean*
ICollection~String~ : +Remove(String item) Boolean*

class IEnumerable
<<interface>> IEnumerable
IEnumerable : +GetEnumerator() IEnumerator*

class IEnumerable~String~
<<interface>> IEnumerable~String~
IEnumerable~String~ : +GetEnumerator() IEnumerator~String~*

class IList
<<interface>> IList
IList : +Add(Object value) Int32*
IList : +Clear() Void*
IList : +Contains(Object value) Boolean*
IList : +get_IsFixedSize() Boolean*
IList : +get_IsReadOnly() Boolean*
IList : +get_Item(Int32 index) Object*
IList : +IndexOf(Object value) Int32*
IList : +Insert(Int32 index, Object value) Void*
IList : +Remove(Object value) Void*
IList : +RemoveAt(Int32 index) Void*
IList : +set_Item(Int32 index, Object value) Void*

class IList~String~
<<interface>> IList~String~
IList~String~ : +get_Item(Int32 index) String*
IList~String~ : +IndexOf(String item) Int32*
IList~String~ : +Insert(Int32 index, String item) Void*
IList~String~ : +RemoveAt(Int32 index) Void*
IList~String~ : +set_Item(Int32 index, String value) Void*

class IReadOnlyCollection~String~
<<interface>> IReadOnlyCollection~String~
IReadOnlyCollection~String~ : +get_Count() Int32*

class IReadOnlyList~String~
<<interface>> IReadOnlyList~String~
IReadOnlyList~String~ : +get_Item(Int32 index) String*

class List~String~
List~String~ : +Add(String item) Void
List~String~ : +AddRange(IEnumerable~String~ collection) Void
List~String~ : -AddWithResize(String item) Void
List~String~ : +AsReadOnly() ReadOnlyCollection~String~
List~String~ : +BinarySearch(Int32 index, Int32 count, String item, IComparer~String~ comparer) Int32
List~String~ : +BinarySearch(String item) Int32
List~String~ : +BinarySearch(String item, IComparer~String~ comparer) Int32
List~String~ : +Clear() Void
List~String~ : +Contains(String item) Boolean
List~String~ : +ConvertAll(Converter~String, TOutput~ converter) List~TOutput~
List~String~ : +CopyTo(String[] array) Void
List~String~ : +CopyTo(Int32 index, String[] array, Int32 arrayIndex, Int32 count) Void
List~String~ : +CopyTo(String[] array, Int32 arrayIndex) Void
List~String~ : +EnsureCapacity(Int32 capacity) Int32
List~String~ : +Exists(Predicate~String~ match) Boolean
List~String~ : +Find(Predicate~String~ match) String
List~String~ : +FindAll(Predicate~String~ match) List~String~
List~String~ : +FindIndex(Predicate~String~ match) Int32
List~String~ : +FindIndex(Int32 startIndex, Predicate~String~ match) Int32
List~String~ : +FindIndex(Int32 startIndex, Int32 count, Predicate~String~ match) Int32
List~String~ : +FindLast(Predicate~String~ match) String
List~String~ : +FindLastIndex(Predicate~String~ match) Int32
List~String~ : +FindLastIndex(Int32 startIndex, Predicate~String~ match) Int32
List~String~ : +FindLastIndex(Int32 startIndex, Int32 count, Predicate~String~ match) Int32
List~String~ : +ForEach(Action~String~ action) Void
List~String~ : +get_Capacity() Int32
List~String~ : +get_Count() Int32
List~String~ : +get_Item(Int32 index) String
List~String~ : +GetEnumerator() Enumerator~String~
List~String~ : +GetRange(Int32 index, Int32 count) List~String~
List~String~ : -Grow(Int32 capacity) Void
List~String~ : +IndexOf(String item) Int32
List~String~ : +IndexOf(String item, Int32 index) Int32
List~String~ : +IndexOf(String item, Int32 index, Int32 count) Int32
List~String~ : +Insert(Int32 index, String item) Void
List~String~ : +InsertRange(Int32 index, IEnumerable~String~ collection) Void
List~String~ : -IsCompatibleObject(Object value) Boolean$
List~String~ : +LastIndexOf(String item) Int32
List~String~ : +LastIndexOf(String item, Int32 index) Int32
List~String~ : +LastIndexOf(String item, Int32 index, Int32 count) Int32
List~String~ : +Remove(String item) Boolean
List~String~ : +RemoveAll(Predicate~String~ match) Int32
List~String~ : +RemoveAt(Int32 index) Void
List~String~ : +RemoveRange(Int32 index, Int32 count) Void
List~String~ : +Reverse() Void
List~String~ : +Reverse(Int32 index, Int32 count) Void
List~String~ : +set_Capacity(Int32 value) Void
List~String~ : +set_Item(Int32 index, String value) Void
List~String~ : +Sort() Void
List~String~ : +Sort(IComparer~String~ comparer) Void
List~String~ : +Sort(Int32 index, Int32 count, IComparer~String~ comparer) Void
List~String~ : +Sort(Comparison~String~ comparison) Void
List~String~ : -System.Collections.Generic.ICollection<T>.get_IsReadOnly() Boolean
List~String~ : -System.Collections.Generic.IEnumerable<T>.GetEnumerator() IEnumerator~String~
List~String~ : -System.Collections.ICollection.CopyTo(Array array, Int32 arrayIndex) Void
List~String~ : -System.Collections.ICollection.get_IsSynchronized() Boolean
List~String~ : -System.Collections.ICollection.get_SyncRoot() Object
List~String~ : -System.Collections.IEnumerable.GetEnumerator() IEnumerator
List~String~ : -System.Collections.IList.Add(Object item) Int32
List~String~ : -System.Collections.IList.Contains(Object item) Boolean
List~String~ : -System.Collections.IList.get_IsFixedSize() Boolean
List~String~ : -System.Collections.IList.get_IsReadOnly() Boolean
List~String~ : -System.Collections.IList.get_Item(Int32 index) Object
List~String~ : -System.Collections.IList.IndexOf(Object item) Int32
List~String~ : -System.Collections.IList.Insert(Int32 index, Object item) Void
List~String~ : -System.Collections.IList.Remove(Object item) Void
List~String~ : -System.Collections.IList.set_Item(Int32 index, Object value) Void
List~String~ : +ToArray() String[]
List~String~ : +TrimExcess() Void
List~String~ : +TrueForAll(Predicate~String~ match) Boolean

class Object
Object : +Equals(Object obj) Boolean
Object : +Equals(Object objA, Object objB) Boolean$
Object : Finalize() Void
Object : +GetHashCode() Int32
Object : +GetType() Type
Object : MemberwiseClone() Object
Object : +ReferenceEquals(Object objA, Object objB) Boolean$
Object : +ToString() String


`, 
                g => {
                    renderTarget.innerHTML = g 
                });
        },
        (error) => {
            console.log(error);
        });
}
// ensure `require` is available globally
if ((typeof(require) !==  typeof(Function)) || (typeof(require.config) !== typeof(Function))) {
    let require_script = document.createElement('script');
    require_script.setAttribute('src', 'https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js');
    require_script.setAttribute('type', 'text/javascript');
    
    
    require_script.onload = function() {
        loadMermaid_00000000000000000000000000000000();
    };

    document.getElementsByTagName('head')[0].appendChild(require_script);
}
else {
    loadMermaid_00000000000000000000000000000000();
}

</script>
<div id="00000000000000000000000000000000"></div>
</div>
